### PR TITLE
Add support for assets in Fuchsia packages

### PR DESCRIPTION
**Pull request by cyngus84**

### CODE REVIEW

**Comment by abarth** [comment]
Looks reasonable to me, but you probably want someone who works on this tool more actively to review.  Maybe @tvolkert or @jason-simmons ?

**Comment by Hixie** [comment]
cc @szakarias for review

File: packages/flutter_tools/lib/src/flx.dart
Comment (tvolkert) [inline packages/flutter_tools/lib/src/flx.dart]: This is unused

File: packages/flutter_tools/bin/fuchsia_asset_builder.dart
Comment (tvolkert) [inline packages/flutter_tools/bin/fuchsia_asset_builder.dart]: Ooph - another copy of this bootstrap code.  Can you refactor to share this between here and `fucshia_builder.dart`?

Comment (tvolkert) [reply packages/flutter_tools/bin/fuchsia_asset_builder.dart]: Perhaps:

    typedef Future<Null> Runner(List<String> args);

    Future<Null> runInContext(List<String> args, Runner runner) {
      ...
    }

File: packages/flutter_tools/bin/fuchsia_asset_builder.dart
Comment (tvolkert) [inline packages/flutter_tools/bin/fuchsia_asset_builder.dart]: May as well make this async so you can avoid the sync file I/O

File: packages/flutter_tools/bin/fuchsia_asset_builder.dart
Comment (tvolkert) [inline packages/flutter_tools/bin/fuchsia_asset_builder.dart]: Nit: this is superfluous

**Review by tvolkert — COMMENTED** [review]
Leaving the review of the asset bundling to @szakarias since she's overhauling that in the Flutter engine.

File: packages/flutter_tools/bin/fuchsia_asset_builder.dart
Comment (cyngus84) [reply packages/flutter_tools/bin/fuchsia_asset_builder.dart]: The synchronous file I/O was on purpose. I'm not sure of the guarantees Dart provides, but I wanted to make sure the files would be written before the program terminates. I also wanted to make sure that the file writes occur in order. Can you help me understand if my concerns are handled by Dart and I should just make this async?

Comment (tvolkert) [reply packages/flutter_tools/bin/fuchsia_asset_builder.dart]: You can just use `await` on the async calls to provide all those guarantees.

File: packages/flutter_tools/lib/src/flx.dart
Comment (cyngus84) [reply packages/flutter_tools/lib/src/flx.dart]: Done

File: packages/flutter_tools/bin/fuchsia_asset_builder.dart
Comment (cyngus84) [reply packages/flutter_tools/bin/fuchsia_asset_builder.dart]: Done

File: packages/flutter_tools/bin/fuchsia_asset_builder.dart
Comment (cyngus84) [reply packages/flutter_tools/bin/fuchsia_asset_builder.dart]: In this case don't I end up awaiting all the async calls in this function and then the function itself in the call to it?

Comment (tvolkert) [reply packages/flutter_tools/bin/fuchsia_asset_builder.dart]: Yep.  It'd be:

```
  ...
  await writeFuchsiaManifest(assets, argResults[_kOptionWorking], outputMan);
}

Future<Null> writeFuchsiaManifest(AssetBundle assets, String outputBase, String fileDest) async {
  syncOp();
  await asyncOp();
  ...
}
```

The benefits are:
* It frees up the main isolate to do other work in the meantime (e.g. animating a "waiting..." cursor if you've set one up).
* It allows you to add a timeout that will cause the tool to fail rather than just hanging forever.  In the main flutter tool, we've found places where file I/O backed by a network file system can hang, and if you're using synchronous operations, your whole tool just hangs.

Comment (cyngus84) [reply packages/flutter_tools/bin/fuchsia_asset_builder.dart]: Ah, okay, thanks for explaining why this is worth doing. I don't really write Dart code so I'm not as aware of these implications as I maybe should be.

File: packages/flutter_tools/bin/fuchsia_asset_builder.dart
Comment (cyngus84) [reply packages/flutter_tools/bin/fuchsia_asset_builder.dart]: Okay, I did something, hopefully it is approximately correct.

File: packages/flutter_tools/bin/fuchsia_asset_builder.dart
Comment (cyngus84) [reply packages/flutter_tools/bin/fuchsia_asset_builder.dart]: Done

**Review by tvolkert — APPROVED** [review]
LGTM.

/cc @szakarias

**Comment by cyngus84** [comment]
Ugh, so the CI is reporting errors like "[warning] Element 'IOSink' from SDK library 'io.dart' is implicitly hidden by 'interface.dart'. (fuchsia_asset_builder.dart, line 84, col 9)" which looks like https://github.com/dart-lang/sdk/issues/28023 which was closed as "seems okay now". Am I doing something wrong, is this a bug, or door number 3?

**Comment by tvolkert** [comment]
Just don't import `dart:io` -- your import of `'../lib/src/base/io.dart'` covers it.

**Comment by cyngus84** [comment]
Waiting for @szakarias to sign off before merging, is that correct?

**Comment by tvolkert** [comment]
Yep - thanks!

**Comment by szakarias** [comment]
LGTM, moving forward will you need both the fuchsia_builder and the fuchsia_asset_builder?

**Comment by cyngus84** [comment]
I believe that once the FLX path for Fuchsia is deprecated fuchsia_builder will no longer be needed.

**Comment by cyngus84** [comment]
Who is the right person to merge this?

**Pull request merged**
